helix-core/src/main/java/org/apache/helix/messaging/handling/HelixTaskExecutor.java

```diff
@@ -714,7 +741,7 @@ private void syncSessionToController(HelixManager manager) {
 
   @Override
   @PreFetch(enabled = false)
-  public void onMessage(String instanceName, List<Message> messages,
+  public synchronized void onMessage(String instanceName, List<Message> messages,
```

Comment:
Why? The periodic refresh does not need to be done synchronized, I think.
If your timer action just adds the refresh event to the queue, there is no need to modify this.